### PR TITLE
custom notebooks are added to the image

### PR DESCRIPTION
**Pull request by ChaamC**
This PR adds the feature for the customization of included tutorial-notebooks.

- The notebooks for the specific images are included when building the images. All notebooks are saved directly in this repo, in the folder related to their image.

This PR relates to the PR in pavics-jupyter-images : https://github.com/bird-house/pavics-jupyter-base/pull/2

### CODE REVIEW

**Review by matprov — APPROVED** [review]

**Review by dbyrns — APPROVED** [review]
LGTM, but we need to take care of https://github.com/bird-house/pavics-jupyter-base/pull/2 before merging this one.

File: eo/Dockerfile
Comment (tlvu) [inline eo/Dockerfile]: I am surprised notebooks are copied directly into the image.  The the other associated PR https://github.com/bird-house/pavics-jupyter-base/pull/2 you just wrote "avoid having to rebuild the image just to update a notebook".

I thought I would see this child image leverage the new feature in the base image.

**Review by tlvu — COMMENTED** [review]
I am surprised but this is CRIM image. I leave it to you guys.

File: eo/Dockerfile
Comment (dbyrns) [reply eo/Dockerfile]: Adding to what I said here : https://github.com/bird-house/pavics-jupyter-base/pull/2#issuecomment-814238353, we looked for an easy way to add notebooks to the ones provide in the base image. We agree that having a custom image with static notebooks is the bare minimum, but it fits our needs right now. We do not expect a lot of modifications in a near future and until we build new images weekly, having another repository just for some notebooks, seems overkill. Adding a wget in this image is in the plan given that the notebooks start to be updated more frequently.

**Comment by ChaamC** [comment]
PR has been updated for a new solution. Please refer to the [other PR](https://github.com/bird-house/birdhouse-deploy/pull/150) for more details.

@MatProv @dbyrns @tlvu

File: eo/Dockerfile
Comment (dbyrns) [inline eo/Dockerfile]: Ok like I said, I can now really see that we use the template as is, so it's not a template. One cannot expect some variables to be replaced like in the birdhouse-deploy config folder.

File: eo/notebook_config.yml.template
Comment (dbyrns) [inline eo/notebook_config.yml.template]: That's very neat. I hadn't thought to reuse the same repo both for the image and notebooks.

File: eo/notebook_config.yml.template
Comment (dbyrns) [inline eo/notebook_config.yml.template]: You see here, what could prevent an image maintainer to cause trouble and set the dest_dir to ${JUPYTERHUB_USER_DATA_DIR}? (See : https://github.com/bird-house/birdhouse-deploy/pull/150#discussion_r616723622)

File: nlp/Dockerfile
Comment (dbyrns) [inline nlp/Dockerfile]: Idem

File: nlp/notebook_config.yml.template
Comment (dbyrns) [inline nlp/notebook_config.yml.template]: Idem

**Review by dbyrns — CHANGES_REQUESTED** [review]
Some requests based on the main PR comments.

**Comment by ChaamC** [comment]
Updated repo : removed dest_dir parameter from the script config

File: eo/notebook_config.yml
Comment (tlvu) [inline eo/notebook_config.yml:8]: To be really safe, use `origin/master` because `master` is the local branch ... which might not always track the proper `origin/master`.

Same comment for the other `nlp/notebook_config.yml` as well.

**Review by tlvu — APPROVED** [review]
+1 for me here, just minor comment.

File: eo/notebook_config.yml
Comment (ChaamC) [reply eo/notebook_config.yml:8]: I think `master` is okay here. The way it works in the script is that it builds a full url using the values in the config here. 
In this code `FULL_URL=$GIT_REPO/branches/$BRANCH/$SOURCE_PATH`, the branch parameter is just inserted in the url to access to the right place on the online repo. No need to distinguish between local or origin since it is a path found on the github site.

Comment (tlvu) [reply eo/notebook_config.yml:8]: Oh yes indeed you are right, my bad.  You do not perform a checkout so there is no local vs remote flavor of `master` to worry about.  I was thinking in term of my script which does a local checkout.  Sorry for the false positive.

Comment (ChaamC) [reply eo/notebook_config.yml:8]: No problem, I understand where it was coming from! :)

**Comment by ChaamC** [comment]
Updated PR : 
- added a dest_sub_dir parameter to specify a subdirectory destination where to download the data. Parameter is optional

**Review by dbyrns — APPROVED** [review]
Last little thing : I would like to see the common repository included here for both eo/nlp and a confirmation that both repos are keep synchronized for each images.

File: eo/CHANGELOG.md
Comment (tlvu) [inline eo/CHANGELOG.md]: finalize the date I guess.

File: eo/notebook_config.yml
Comment (tlvu) [inline eo/notebook_config.yml]: I would expect `dest_sub_dir: crim-eo`?  But I am not sure.

**Review by tlvu — APPROVED** [review]
LGTM.  A remark about `dest_sub_dir` but I am not sure.

File: eo/notebook_config.yml
Comment (dbyrns) [reply eo/notebook_config.yml]: @tlvu see my comment here : https://github.com/bird-house/pavics-jupyter-base/pull/3#discussion_r625983632
Like this, the notebooks would be checkout to : /data/jupyterhub_user_data/tutorial-notebooks/crim-eo
What I asked to @ChaamC would make it clearer to show how to achieve the following strucutre: 
/data/jupyterhub_user_data/tutorial-notebooks/crim-eo/eo
/data/jupyterhub_user_data/tutorial-notebooks/crim-eo/common

Comment (ChaamC) [reply eo/notebook_config.yml]: I can add the suggested structure. 
But what would be the repo_url/branch/source_path for the common folder?
Should I try to find all the notebooks that are found in the tutorial-notebooks folder (when spawning the pavics-workflow image for example)?
Or maybe I can just add this [folder](https://github.com/Ouranosinc/pavics-sdi/tree/master/docs/source/notebooks) for now?

Comment (dbyrns) [reply eo/notebook_config.yml]: I thought that the current tutorial-notebooks folder was filled using only that folder. But looking at the diff, this is enough for our common folder. Maybe we could call it "core"?

Comment (ChaamC) [reply eo/notebook_config.yml]: Yes, there are more in the current tutorial-notebooks folder, so more sources are necessary if we want them all. But I don't think all those notebooks will work anyway. The eo image for example has only the base packages and the specific packages for eo. The packages necessary  for the current tutorial-notebooks folder are not all on the eo image.

Even with the linked folder in my previous comment, I am not sure which notebooks will work or not from that list.

So, do I just take the linked folder for now, and store it in a 'core' folder, for both eo/nlp?

Comment (dbyrns) [reply eo/notebook_config.yml]: Yes go for it. At least for people using those images, they will see what PAVICS has to offer and the mecanism will be demonstrated. If some libraries are missing we will fix that later.

Comment (tlvu) [reply eo/notebook_config.yml]: FYI info if you guys want to replicate exactly what in the current `tutorial-notebooks` folder into your `common` folder https://github.com/bird-house/birdhouse-deploy/blob/20d440fd293b01d4cf3e1dccf2933877a651d571/birdhouse/deployment/trigger-deploy-notebook#L60-L71

However, keep in mind that bringing in notebook that your image can not run (missing packages) is no cool !   I'd suggest keeping only the notebooks relevant to your image runtime capabilities.

With that in mind, pavics-sdi is a good start (probably most notebooks will run without missing packages).  Finch and Raven notebooks will requires more packages to be installed.

Comment (dbyrns) [reply eo/notebook_config.yml]: > However, keep in mind that bringing in notebook that your image can not run (missing packages) is no cool !

I agree, but it's more as a place holder or demonstration capabilities right now. I think that once this setup is ready we will be able to better partitionize the notebooks for all our (Ouranos and CRIM) images.

Comment (ChaamC) [reply eo/notebook_config.yml]: Common folder has been added to configs :)

File: nlp/notebook_config.yml
Comment (tlvu) [inline nlp/notebook_config.yml:12]: Ouch double download of the pavics-sdi notebooks.  At least the config is clearer.

We could have one `config.yml` that will handle all the notebooks repos (eo, nlp, common) but then each image won't be "independent" anymore.

Don't have a quick solution at the moment.

Comment (dbyrns) [reply nlp/notebook_config.yml:12]: @tlvu, at least it's one copy per image and not per user. Hopefully, the common notebooks folder will stay small.

> We could have one config.yml that will handle all the notebooks repos (eo, nlp, common) but then each image won't be "independent" anymore.

That's the whole point of this PR! Having a single config.yml is about the same as what we have right now. 

The download script could be optimized (not for this PR) and keep a cache of each repo independently of the current image. So that while looping over the images, if a repo has already been updated, it uses the cache. 

But again I'm note sure of your concern, is it the download or the disk space? In both case it looks like trying to save pennies.

Comment (tlvu) [reply nlp/notebook_config.yml:12]: Opps sorry, sending wrong impression again.  Wasn't saying this is a deal breaker, else I would change the approval status.

Probably the layout on disk could be
```
/data/jupyterhub_user_data/tutorial-notebooks/common
/data/jupyterhub_user_data/tutorial-notebooks/crim-eo
/data/jupyterhub_user_data/tutorial-notebooks/crim-nlp
```
Instead of
```
/data/jupyterhub_user_data/tutorial-notebooks/crim-eo/eo
/data/jupyterhub_user_data/tutorial-notebooks/crim-eo/common
/data/jupyterhub_user_data/tutorial-notebooks/crim-nlp/nlp
/data/jupyterhub_user_data/tutorial-notebooks/crim-nlp/common
```

But that also assume the same definition of `common` for those images, which might make sense only if on the same deployment node.  Like you said, it's pennies at this point so let's do this once we get there.

Comment (dbyrns) [reply nlp/notebook_config.yml:12]: Don't worry I'm not offended and sorry if this is the impression I leave.

Comment (tlvu) [reply nlp/notebook_config.yml:12]: Just to be clear, I like all trade-offs/limitations to be clearly documented in a PR.  Doesn't mean we need to address all of them immediately.  I much prefer iterative approach where we improve gradually than keep adding feature to grow the PR and never release it!

**Pull request merged**
